File: tests/cpp/src/BasicResponder.cpp
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

#include <cameo/api/cameo.h>
#include <iostream>

using namespace std;
using namespace cameo;

int main(int argc, char *argv[]) {

	This::init(argc, argv);

	int options = 0;
	bool useProxy = false;
	if (argc > 2) {
		useProxy = (string(argv[1]) == "true");
	}
	if (useProxy) {
		options |= option::USE_PROXY;
	}

	unique_ptr<coms::basic::Responder> responder;

	try {
		cout << "Creating responder" << endl;

		responder = coms::basic::Responder::create("responder");
		responder->init();
	}
	catch (const InitException& e) {
		cout << "Responder error" << endl;
		return -1;
	}

	cout << "Created responder " << *responder << endl;

	This::setRunning();

	// Receive first request.
	unique_ptr<coms::basic::Request> request = responder->receive();
	cout << "Received request " << *request << endl;

	request->reply("1st response");
	request->reply("1st response (bis)");

	// Receive second request.
	request = responder->receive();

	cout << "Received request " << request->get() << " " << request->getSecondPart() << endl;
	request->reply("2nd response");

	// Receive third request.
	request = responder->receive();
	cout << "Received request " << request->get() << endl;

	// Reply.
	request->reply("3rd response");
	cout << "Replied 3rd response" << endl;

	// Receive request.
	request = responder->receive();
	cout << "received request " << *request << endl;

	cout << "Wait so that the requester has timed out" << endl;
	this_thread::sleep_for(chrono::seconds(1));

	request->reply("4th response");


	// Receive request.
	request = responder->receive();
	cout << "received request " << *request << endl;
	request->reply("5th response");


	std::unique_ptr<ServerAndApp> requester = request->connectToRequester(options);
	cout << "Requester " << requester->getApp() << endl;


	cout << "Finished the application" << endl;

	return 0;
}

File: tests/python/src/basicresponder.py
import sys
import time
import cameopy

this = cameopy.This
this.init(sys.argv)

useProxy = False if len(sys.argv) < 2 else (sys.argv[1] == "true")

options = 0
if useProxy:
    options |= cameopy.option.USE_PROXY

print("Creating responder")

responder = cameopy.coms.basic.Responder.create("responder")
responder.init()

this.setRunning()

# Receive first request.
request = responder.receive();
print("Received request", request.getString())

request.reply("1st response")
request.reply("1st response (bis)")

# Receive second request.
request = responder.receive();

sp = request.getSecondPart()

print("Received request with parts", request.getString(), request.getSecondPart().decode("utf-8"))

request.reply("2nd response")

#if not res:
#    print("Error, no timeout expected with", request)    


# Receive third request.
request = responder.receive();
print("Received request", request.getString())

request.reply("3rd response")

# Receive the fourth request.
request = responder.receive();
print("Received request", request.getString())

print("Wait so that the requester has timed out")

time.sleep(1)

request.reply("4th response")

# Receive the fifth request.
request = responder.receive();
print("Received request " + request.getString())

# Reply.
request.reply("5th response")

print("Replied 5th")

# Test connection.
requester = request.connectToRequester(options)
app = requester.getApp()

print("Requester", app.getId())

print("Finished the application")
